<a href="https://colab.research.google.com/github/EltonGS/BC_DS_projeto_modulo_3/blob/main/Notebooks/tratamento_dados.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Informações sobre a base de dados

Ao verificar a página de onde os dados foram extraídos, [Brasil.IO](https://brasil.io/dataset/covid19/caso_full/), temos a informaçãos que alguns dados podem ser repetidos no dataset quando os dados não foram divulgados pela **SES - Secretarias Estaduais de Saúde**.

No dataset temos as colunas **is_last** e **is_repeated** que identificam quando um dado é a **última atualização disponível** e quando o **dado é repetido**, respectivamente.

In [90]:
# importando as bibliotecas que serão usadas
import pandas as pd
import numpy as np

In [91]:
# caminho de onde estão os dados utilizados
url = '/content/drive/MyDrive/Colab Notebooks/Bootcamp Data Science/Módulo 3/Arquivos/caso_full.zip'

# lendo o arquivo com os dados que estão comprimidos no padrão zip
dados_totais = pd.read_csv(url, compression='zip')

In [92]:
# visualizando os dados
dados_totais.head(3)

city  city_ibge_code        date  ...  state  new_confirmed  new_deaths
0  Rio Branco       1200401.0  2020-03-17  ...     AC              3           0
1         NaN            12.0  2020-03-17  ...     AC              3           0
2  Rio Branco       1200401.0  2020-03-18  ...     AC              0           0

[3 rows x 18 columns]

### É possível perceber uma linha abaixo dos dados da cidade de Rio Branco que parece repetir o dado da linha acima, ao repararmos na coluna place_type vemos que essa segunda linha se refere ao estado onde a cidade está. Para a cidade de Rio Branco temos os dados do estado do Acre.

### Para esse trabalho focaremos na cidade de Santo André no estado de São Paulo, onde resido.

In [93]:
# verificando os dados que são da cidade (city) santo andré E do estado (state) SP
dados_sa = dados_totais.query("city == 'Santo André' & state == 'SP'")
dados_sa

city  city_ibge_code  ... new_confirmed  new_deaths
1995526  Santo André       3547809.0  ...             1           0
1995536  Santo André       3547809.0  ...             0           0
1995551  Santo André       3547809.0  ...             5           0
1995569  Santo André       3547809.0  ...            -4           0
1995592  Santo André       3547809.0  ...             1           0
...              ...             ...  ...           ...         ...
2265700  Santo André       3547809.0  ...           211          10
2266347  Santo André       3547809.0  ...           331           8
2266994  Santo André       3547809.0  ...           201           5
2267641  Santo André       3547809.0  ...            93           6
2268288  Santo André       3547809.0  ...             0           0

[470 rows x 18 columns]

Nos dados mostrados acima temos 470 linhas para a cidade Santo André em SP, e temos também o código IBGE da cidade: **3547809.0**

Assim podemos verificar se temos todos os dados referente a essa cidade.

In [94]:
# extração dos dados pelo código da cidade
dados_sa2 = dados_totais.query("city_ibge_code == 3547809.0")
dados_sa2

city  city_ibge_code  ... new_confirmed  new_deaths
1995526  Santo André       3547809.0  ...             1           0
1995536  Santo André       3547809.0  ...             0           0
1995551  Santo André       3547809.0  ...             5           0
1995569  Santo André       3547809.0  ...            -4           0
1995592  Santo André       3547809.0  ...             1           0
...              ...             ...  ...           ...         ...
2265700  Santo André       3547809.0  ...           211          10
2266347  Santo André       3547809.0  ...           331           8
2266994  Santo André       3547809.0  ...           201           5
2267641  Santo André       3547809.0  ...            93           6
2268288  Santo André       3547809.0  ...             0           0

[470 rows x 18 columns]

Os dados extraídos das duas maneiras são os mesmos. Continuaremos com os dados do dataframe **dados_sa**.

In [95]:
# verificação se há alguma dado NaN em nosso dataframe
dados_sa.isna().value_counts()

city   city_ibge_code  date   epidemiological_week  estimated_population  estimated_population_2019  is_last  is_repeated  last_available_confirmed  last_available_confirmed_per_100k_inhabitants  last_available_date  last_available_death_rate  last_available_deaths  order_for_place  place_type  state  new_confirmed  new_deaths
False  False           False  False                 False                 False                      False    False        False                     False                                          False                False                      False                  False            False       False  False          False         470
dtype: int64

In [96]:
# redefinindo o index para facilitar o trabalho com os dados
dados_sa = dados_sa.reset_index()

# dados com o index resetado
dados_sa.head(3)

index         city  city_ibge_code  ... state  new_confirmed  new_deaths
0  1995526  Santo André       3547809.0  ...    SP              1           0
1  1995536  Santo André       3547809.0  ...    SP              0           0
2  1995551  Santo André       3547809.0  ...    SP              5           0

[3 rows x 19 columns]

In [97]:
# verificando se temos alguma informação True na coluna is_last
dados_sa['is_last'].value_counts()

False    469
True       1
Name: is_last, dtype: int64

In [98]:
# verificando onde está o dado True identificado acima
dados_sa.query("is_last == True")

index         city  city_ibge_code  ... state  new_confirmed  new_deaths
468  2267641  Santo André       3547809.0  ...    SP             93           6

[1 rows x 19 columns]

In [99]:
# verificando os dados da linha 468 do index
dados_sa.loc[468]

index                                                2267641
city                                             Santo André
city_ibge_code                                   3.54781e+06
date                                              2021-06-27
epidemiological_week                                  202126
estimated_population                                  721368
estimated_population_2019                             718773
is_last                                                 True
is_repeated                                            False
last_available_confirmed                               57811
last_available_confirmed_per_100k_inhabitants        8014.08
last_available_date                               2021-06-27
last_available_death_rate                             0.0476
last_available_deaths                                   2750
order_for_place                                          469
place_type                                              city
state                   

Podemos ver na informação acima que a **última atualização disponível** para **Santo André / SP** é da data **27/06/2021**.

Agora vamos verificar se em nosso dataframe temos algum dado repetido.

In [100]:
# verificando se temos alguma informação True na coluna is_repeated
dados_sa['is_repeated'].value_counts()

False    442
True      28
Name: is_repeated, dtype: int64

In [101]:
# verificando quais são os dados repetidos que temos no datafarme
dados_repetidos = dados_sa.query("is_repeated == True")
dados_repetidos

index         city  city_ibge_code  ... state  new_confirmed  new_deaths
5    1995615  Santo André       3547809.0  ...    SP              0           0
6    1995638  Santo André       3547809.0  ...    SP              0           0
7    1995661  Santo André       3547809.0  ...    SP              0           0
8    1995684  Santo André       3547809.0  ...    SP              0           0
12   1995830  Santo André       3547809.0  ...    SP              0           0
13   1995870  Santo André       3547809.0  ...    SP              0           0
166  2072247  Santo André       3547809.0  ...    SP              0           0
175  2078070  Santo André       3547809.0  ...    SP              0           0
202  2095539  Santo André       3547809.0  ...    SP              0           0
203  2096186  Santo André       3547809.0  ...    SP              0           0
205  2097480  Santo André       3547809.0  ...    SP              0           0
214  2103303  Santo André       3547809.0  ...    SP              0           0
230  2113655  Santo André       3547809.0  ...    SP              0           0
235  2116890  Santo André       3547809.0  ...    SP              0           0
236  2117537  Santo André       3547809.0  ...    SP              0           0
237  2118184  Santo André       3547809.0  ...    SP              0           0
238  2118831  Santo André       3547809.0  ...    SP              0           0
249  2125948  Santo André       3547809.0  ...    SP              0           0
275  2142770  Santo André       3547809.0  ...    SP              0           0
289  2151828  Santo André       3547809.0  ...    SP              0           0
291  2153122  Santo André       3547809.0  ...    SP              0           0
404  2226233  Santo André       3547809.0  ...    SP              0           0
411  2230762  Santo André       3547809.0  ...    SP              0           0
426  2240467  Santo André       3547809.0  ...    SP              0           0
442  2250819  Santo André       3547809.0  ...    SP              0           0
455  2259230  Santo André       3547809.0  ...    SP              0           0
458  2261171  Santo André       3547809.0  ...    SP              0           0
469  2268288  Santo André       3547809.0  ...    SP              0           0

[28 rows x 19 columns]

No dataframe acima (**dados_repetidos**) temos as datas em que os dados foram repetidos por não ter informação divulgada pelo boletime das Secretarias Estaduais de Saúde.

Temos um total de **28** datas.

In [102]:
dados_sa.head(5)

index         city  city_ibge_code  ... state  new_confirmed  new_deaths
0  1995526  Santo André       3547809.0  ...    SP              1           0
1  1995536  Santo André       3547809.0  ...    SP              0           0
2  1995551  Santo André       3547809.0  ...    SP              5           0
3  1995569  Santo André       3547809.0  ...    SP             -4           0
4  1995592  Santo André       3547809.0  ...    SP              1           0

[5 rows x 19 columns]

In [103]:
# renomeando as colunas para facilitar o trabalho com os dados
dados_sa.rename(columns={
    'city' : 'cidade',
    'city_ibge_code' : 'codigo_cidade',
    'date' : 'data',
    'epidemiological_week' : 'semana_epidem',
    'estimated_population' : 'populacao',
    'estimated_population_2019' : 'populacao_2019',
    'last_available_confirmed' : 'casos_confirmados',
    'last_available_confirmed_per_100k_inhabitants' : 'casos_por_100k_habitantes',
    'last_available_date' : 'data_atualizacao',
    'last_available_death_rate' : 'taxa_mortalidade',
    'last_available_deaths' : 'obitos',
    'state' : 'estado',
    'new_confirmed' : 'casos_novos',
    'new_deaths' : 'obitos_novos'}, inplace=True)

In [104]:
# visualizando os novos nomes das colunas
dados_sa.head(3)

index       cidade  codigo_cidade  ... estado  casos_novos  obitos_novos
0  1995526  Santo André      3547809.0  ...     SP            1             0
1  1995536  Santo André      3547809.0  ...     SP            0             0
2  1995551  Santo André      3547809.0  ...     SP            5             0

[3 rows x 19 columns]

In [105]:
# removendo as colunas que não interessam para o projeto
dados_sa.drop(['index', 'order_for_place', 'place_type'], axis=1, inplace=True)

In [106]:
dados_sa.head(3)

cidade  codigo_cidade        data  ...  estado  casos_novos  obitos_novos
0  Santo André      3547809.0  2020-03-16  ...      SP            1             0
1  Santo André      3547809.0  2020-03-17  ...      SP            0             0
2  Santo André      3547809.0  2020-03-18  ...      SP            5             0

[3 rows x 16 columns]

Agora temos os dados para **Santo André / SP** tratados.

Foram identificadas as linhas com dados repetidos, se haviam dados NaN, o index foi resetados, as colunas foram renomeadas e as colunas que não interessam para essa análise foram removidas.

Para efeito de comparação com os dados de Santo André, também serão extraídos os dados do estado de São Paulo.


In [107]:
# extraindo a informação para o estado de SP
dados_estadual_sp = dados_totais.query("state == 'SP' & place_type == 'state'")

In [108]:
# visualizando os dados para o estado de SP
dados_estadual_sp

city  city_ibge_code        date  ...  state  new_confirmed  new_deaths
1995468  NaN            35.0  2020-02-25  ...     SP              1           0
1995470  NaN            35.0  2020-02-26  ...     SP              0           0
1995472  NaN            35.0  2020-02-27  ...     SP              0           0
1995474  NaN            35.0  2020-02-28  ...     SP              1           0
1995476  NaN            35.0  2020-02-29  ...     SP              0           0
...      ...             ...         ...  ...    ...            ...         ...
2265817  NaN            35.0  2021-06-24  ...     SP          17951         781
2266464  NaN            35.0  2021-06-25  ...     SP          20994         772
2267111  NaN            35.0  2021-06-26  ...     SP          18932         538
2267758  NaN            35.0  2021-06-27  ...     SP           7965         134
2268405  NaN            35.0  2021-06-28  ...     SP           4285          62

[490 rows x 18 columns]

Temos um dataframe com **490 linhas** e **18 colunas**.

In [109]:
# extraindo os dados pelo código do IBGE para comparar
dados_estadual_sp2 = dados_totais.query("city_ibge_code == 35.0")

# visualizando os dados extraídos
dados_estadual_sp2

city  city_ibge_code        date  ...  state  new_confirmed  new_deaths
1995468  NaN            35.0  2020-02-25  ...     SP              1           0
1995470  NaN            35.0  2020-02-26  ...     SP              0           0
1995472  NaN            35.0  2020-02-27  ...     SP              0           0
1995474  NaN            35.0  2020-02-28  ...     SP              1           0
1995476  NaN            35.0  2020-02-29  ...     SP              0           0
...      ...             ...         ...  ...    ...            ...         ...
2265817  NaN            35.0  2021-06-24  ...     SP          17951         781
2266464  NaN            35.0  2021-06-25  ...     SP          20994         772
2267111  NaN            35.0  2021-06-26  ...     SP          18932         538
2267758  NaN            35.0  2021-06-27  ...     SP           7965         134
2268405  NaN            35.0  2021-06-28  ...     SP           4285          62

[490 rows x 18 columns]

A informação é igual a que foi extraída anteriormente. A partir daqui continuaremos com o dataframe **dados_estadual_sp**.

In [110]:
# verificando se temos dados NaN em nosso dataframe
dados_estadual_sp.isna().value_counts()

city  city_ibge_code  date   epidemiological_week  estimated_population  estimated_population_2019  is_last  is_repeated  last_available_confirmed  last_available_confirmed_per_100k_inhabitants  last_available_date  last_available_death_rate  last_available_deaths  order_for_place  place_type  state  new_confirmed  new_deaths
True  False           False  False                 False                 False                      False    False        False                     False                                          False                False                      False                  False            False       False  False          False         490
dtype: int64

Na coluna **city** temos dados NaN pois os dados não são referentes a nenhuma cidade mas sim ao estado de São Paulo.

Para tornar o trabalho com os dados mais intuitivo e simples, substituiremos a informação NaN pelo nome do estado São Paulo.

In [111]:
# substituindo a informação NaN na coluna city pelo nome do estado São Paulo
dados_estadual_sp['city'].replace(np.nan, 'São Paulo', inplace=True)

# visualizando os dados com a alteração na coluna city
dados_estadual_sp.head(3)

/usr/local/lib/python3.7/dist-packages/pandas/core/series.py:4582: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  method=method,


city  city_ibge_code        date  ...  state  new_confirmed  new_deaths
1995468  São Paulo            35.0  2020-02-25  ...     SP              1           0
1995470  São Paulo            35.0  2020-02-26  ...     SP              0           0
1995472  São Paulo            35.0  2020-02-27  ...     SP              0           0

[3 rows x 18 columns]

In [112]:
# verificando se há mais algum dados NaN no dataframe
dados_estadual_sp.isna().value_counts()

city   city_ibge_code  date   epidemiological_week  estimated_population  estimated_population_2019  is_last  is_repeated  last_available_confirmed  last_available_confirmed_per_100k_inhabitants  last_available_date  last_available_death_rate  last_available_deaths  order_for_place  place_type  state  new_confirmed  new_deaths
False  False           False  False                 False                 False                      False    False        False                     False                                          False                False                      False                  False            False       False  False          False         490
dtype: int64

In [113]:
# resetando o index para facilitar o trabalho com os dados
dados_estadual_sp = dados_estadual_sp.reset_index()

# visualizando os dados com o novo index
dados_estadual_sp.head(3)

index       city  city_ibge_code  ... state  new_confirmed  new_deaths
0  1995468  São Paulo            35.0  ...    SP              1           0
1  1995470  São Paulo            35.0  ...    SP              0           0
2  1995472  São Paulo            35.0  ...    SP              0           0

[3 rows x 19 columns]

In [114]:
# verificando a coluna is_last se temos algum dado True
dados_estadual_sp.query("is_last == True")

index       city  city_ibge_code  ... state  new_confirmed  new_deaths
489  2268405  São Paulo            35.0  ...    SP           4285          62

[1 rows x 19 columns]

In [115]:
dados_estadual_sp.tail(3)

index       city  city_ibge_code  ... state  new_confirmed  new_deaths
487  2267111  São Paulo            35.0  ...    SP          18932         538
488  2267758  São Paulo            35.0  ...    SP           7965         134
489  2268405  São Paulo            35.0  ...    SP           4285          62

[3 rows x 19 columns]

Podemos ver que a informação disponibilizada para o dia 28/06/2021 é a última disponibilizada para o estado de SP e é a última linha no dataframe.

In [116]:
# verificando a coluna is_repeated se temos algum dado True
dados_repetidos_sp = dados_estadual_sp.query("is_repeated == True")

In [117]:
# visualizando os dados com True na coluna is_repeated
dados_repetidos_sp

index       city  city_ibge_code  ... state  new_confirmed  new_deaths
17   1995511  São Paulo            35.0  ...    SP              0           0
19   1995521  São Paulo            35.0  ...    SP              0           0
255  2117007  São Paulo            35.0  ...    SP              0           0
256  2117654  São Paulo            35.0  ...    SP              0           0
257  2118301  São Paulo            35.0  ...    SP              0           0
258  2118948  São Paulo            35.0  ...    SP              0           0
295  2142887  São Paulo            35.0  ...    SP              0           0

[7 rows x 19 columns]

No dataframe acima (**dados_repetidos_sp**) temos as datas em que os dados foram repetidos por não ter informação divulgada pelo boletim das Secretarias Estaduais de Saúde.

Temos um total de **7** datas.

In [118]:
# renomeando as colunas para facilitar o trabalho com os dados
dados_estadual_sp.rename(columns={
    'city' : 'estado',
    'city_ibge_code' : 'codigo_estado',
    'date' : 'data',
    'epidemiological_week' : 'semana_epidem',
    'estimated_population' : 'populacao',
    'estimated_population_2019' : 'populacao_2019',
    'last_available_confirmed' : 'casos_confirmados',
    'last_available_confirmed_per_100k_inhabitants' : 'casos_por_100k_habitantes',
    'last_available_date' : 'data_atualizacao',
    'last_available_death_rate' : 'taxa_mortalidade',
    'last_available_deaths' : 'obitos',
    'state' : 'estado',
    'new_confirmed' : 'casos_novos',
    'new_deaths' : 'obitos_novos'}, inplace=True)

In [119]:
# visualizando as colunas renomeadas
dados_estadual_sp.head(3)

index     estado  codigo_estado  ... estado  casos_novos  obitos_novos
0  1995468  São Paulo           35.0  ...     SP            1             0
1  1995470  São Paulo           35.0  ...     SP            0             0
2  1995472  São Paulo           35.0  ...     SP            0             0

[3 rows x 19 columns]

In [121]:
# removendo as colunas que não interessam para o projeto
dados_estadual_sp.drop(['index', 'order_for_place', 'place_type'], axis=1, inplace=True)

In [122]:
# visualizando o dataframe com as colunas que interessam para o projeto
dados_estadual_sp.head(3)

estado  codigo_estado        data  ...  estado  casos_novos  obitos_novos
0  São Paulo           35.0  2020-02-25  ...      SP            1             0
1  São Paulo           35.0  2020-02-26  ...      SP            0             0
2  São Paulo           35.0  2020-02-27  ...      SP            0             0

[3 rows x 16 columns]

Agora temos os dados para o estado de **São Paulo** tratados.

Foram identificadas as linhas com dados repetidos, se haviam dados NaN, os dados NaN identificados foram substituídos por São Paulo, o index foi resetado, as colunas foram renomeadas e as colunas que não interessam para essa análise foram removidas.

Agora os dataframes que serão utilizados na análise explaratória serão exportados como arquivos CSV.

In [123]:
# exportando o dataframe com os dados de Santo André como arquivo CSV
dados_sa.to_csv('dados_sa.csv', index=False)

# exportando o dataframe com os dados do estado de São Paulo como arquivo CSV
dados_estadual_sp.to_csv('dados_estadual_sp.csv', index=False)

# exportando o dataframe com os dados repetidos de Santo André como arquivo CSV
dados_repetidos.to_csv('dados_repetidos_sa.csv', index=False)

# exportando o dataframe com os dados repetidos do estado de SP como arquivo CSV
dados_repetidos_sp.to_csv('dados_repetidos_sp.csv', index=False)